core/src/main/java/org/renjin/primitives/S3.java

```diff
@@ -259,6 +261,209 @@ public static SEXP tryDispatchFromPrimitive(Context context, Environment rho, Fu
+    String functionEnvName = ".__T__" + functionName + ":base";
+    Environment functionEnv = (Environment) context.getGlobalEnvironment().findVariable(context, Symbol.get(functionEnvName));
+    SEXP function = null;
+    PairList.Builder allArgs = new PairList.Builder();
+    // S4 methods for each generic function is stored in an environment. methods for each signature is stored
+    // separately using the signature as name. for example
+    // setMethod("[", signature("AA","BB","CC"), function(x, i, j, ...))
+    // is stored as `AA#BB#CC` in an environment named `.__T__[:base`
+    // here we get the first method from the method environment and split the name by # to know what the expected
+    // signature length is. This might be longer the length of arguments and #ANY should be used for missing
+    // arguments. In case signature is shorted than the number of arguments we don't need to evaluate the extra
+    // arguments.
+    int signatureLength = ((Symbol) functionEnv.getFrame().getSymbols().toArray()[0]).getPrintName().split("#").length;
+
+    Object[][] allPossibleSignatures = generateAllPossibleSignatures(context, rho, args, null, signatureLength, 0);
+
+    //function = findFunctionMatchingAnyOfSignatures(context, functionEnv, allPossibleSignatures, 0, Symbol.UNBOUND_VALUE);
+
+    Map<Double, SEXP> methods = findFunctionMatchingAnyOfSignatures(context, functionEnv, allPossibleSignatures, 0, null);
+
+    if(methods.isEmpty()) {
+      throw new EvalException("object of type 'S4' is not subsettable");
+    }
+
+    TreeMap<Double, SEXP> map = new TreeMap<>(methods);
+    function = map.get(map.firstKey());
+
+    allArgs.add(source);
+    allArgs.add(args.getElementAsSEXP(1));
+    SEXP result = context.evaluate(new FunctionCall(function, allArgs.build()));
+    return result;
+  }
+
+  private static Map<Double, SEXP> findFunctionMatchingAnyOfSignatures(Context context, Environment functionEnv, Object[][] signatures, int useSignature, Map<Double, SEXP> methods) {
+    // recursively go through arguments and get their class and lookup until the signature is found.
+    // Issues:
+    // - Stops with the first hit
+    // - Is not aware of distance between classes of input arguments to different method signatures
+    if(methods == null) {
+      methods = new HashMap<>(signatures.length);
+    }
+    if(useSignature < signatures.length) {
+      Object[] currentSig = signatures[useSignature];
+      int currentSigLength = currentSig.length;
+      Symbol methodName = Symbol.get((String) currentSig[0]);
+      SEXP function = functionEnv.findVariable(context, methodName);
+
+      if(function != Symbol.UNBOUND_VALUE) {
+        int[] distances = new int[currentSigLength - 1];
+        for(int i = 0; i < currentSigLength - 1; i++) {
+          distances[i] = (Integer) currentSig[i + 1];
+        }
+
+        double methodRating = computeMethodRank(distances);
+        methods.put(methodRating, function);
```

Comment:
If multiple functions have the same methodRank, only the last one will be saved in `methods`. Is the intended behavior?